### PR TITLE
Bulk Attachment Download Frontend

### LINKED ISSUE #16: Bulk download attachments on Storage mode
**Problem**
Currently users have to individually download attachments on the corresponding response view in FormSG. The exported csv shows filename. If users receive a high volume of responses and every one of them has an attachment, individually downloading attachments is going to be a pain in the ass.

How do we make attachment downloads more convenient for users, and best support whatever they need to do with attachments after?

**Why users have to download attachments rather than just viewing them**

For common use cases like applications, users often need to bring up attachments again or send them to other parties for downstream stages of the process.
e.g. need to bring up CV and personal statement again if applicant is shortlisted through later rounds

**Immediate implementation**

summary: https://docs.google.com/document/d/1nzUHQDzo4WBA67xdy0a3A3nJLCeKVSzjzPtJebtS9xQ/edit#heading=h.ypu06omk303h

1. Previous discussions were halted because downloading all attachments within x date range in 1 zip file was not technically feasible. 

2. Alternative was to have browser download 1 zip file (per response) at a time, but would not be a great UX, as user has to click twice (unzip, then open folder) just to access each response's attachments. 

Do we have a better option than option 2?

**Screens**
https://app.zeplin.io/project/5b29ffde48eaaf6318d01b05/dashboard?seid=5f87c5ed5cf4fc4b800abbde

**Longer term**
- Look at Zendesk extension
- Explore service worker and saving to disk

### PR DESCRIPTION
**Pull request by frankchn**
## Problem
This pull request adds the modal and other frontend features necessary to enable bulk attachment downloading.

## Solution
This adds a dropdown and modal and otherwise uses the existing csvDownload and attachment download backends.

Closes #16

### CODE REVIEW

**Review by liangyuanruo — COMMENTED** [review]
code looks great, let me test this on staging later

**Comment by frankchn** [comment]
Added P-Queue based implementation to restrict concurrent downloads to 1:

image

File: src/public/modules/forms/helpers/decryption.worker.js [hunk 1]
```diff
@@ -150,10 +152,10 @@ async function decryptIntoCsv(data) {
           }
         })
 
-        downloadBlob = await downloadAndDecryptAttachmentsAsZip(
-          attachmentDownloadUrls,
-          secretKey,
+        let downloadPromise = queue.add(() =>
+          downloadAndDecryptAttachmentsAsZip(attachmentDownloadUrls, secretKey),
         )
+        downloadBlob = await downloadPromise
```
Comment (liangyuanruo) [inline src/public/modules/forms/helpers/decryption.worker.js:158]:
```suggestion
        downloadBlob = await queue.add(() =>
          downloadAndDecryptAttachmentsAsZip(attachmentDownloadUrls, secretKey),
        )
```

File: src/public/modules/forms/services/submissions.client.factory.js [hunk 2]
```diff
@@ -270,7 +270,12 @@ function SubmissionsFactory(
           let receivedRecordCount = 0
 
           // Create a pool of decryption workers
-          const numWorkers = $window.navigator.hardwareConcurrency || 4
+          let numWorkers = $window.navigator.hardwareConcurrency || 4
+          if (downloadAttachments) {
+            // If we are downloading attachments, we restrict the number of threads
+            // to one to limit resource usage on the client's browser.
+            numWorkers = 1
+          }
```
Comment (liangyuanruo) [inline src/public/modules/forms/services/submissions.client.factory.js:278]:
```suggestion
          // If we are downloading attachments, we restrict the number of threads
          // to one to limit resource usage on the client's browser.
          const numWorkers = downloadAttachments ? 1 : ($window.navigator.hardwareConcurrency || 4)
```

**Review by liangyuanruo — APPROVED** [review]
lgtm - do we have any more UI changes to make?

File: src/public/modules/forms/helpers/decryption.worker.js [hunk 3]
```diff
@@ -150,9 +152,8 @@ async function decryptIntoCsv(data) {
           }
         })
 
-        downloadBlob = await downloadAndDecryptAttachmentsAsZip(
-          attachmentDownloadUrls,
-          secretKey,
+        downloadBlob = await queue.add(() =>
+          downloadAndDecryptAttachmentsAsZip(attachmentDownloadUrls, secretKey),
```
Comment (liangyuanruo) [inline src/public/modules/forms/helpers/decryption.worker.js:156]: to add try-catch block to isolate attachment failure from CSV decryption failure

**Review by liangyuanruo — CHANGES_REQUESTED** [review]
**Action items**
1. To display separate wording in modal if there are no attachments to download.
2. To improve messaging on the modal to emphasiSe that it is beta, experimental, and requires the user to have ample disk space and be mentally prepared for a large number of files to be triggered.
3. To add Google Analytics monitoring (#762)

**Comment by syan-syan** [comment]
Chrome
- pre-download modal should be page center not flushed to the top
- Export dropdown options should not disconnect when user scrolls down

File: src/public/modules/forms/helpers/decryption.worker.js (hunk 3 above)
Comment (frankchn) [reply src/public/modules/forms/helpers/decryption.worker.js]: Added

**Comment by frankchn** [comment]
- Added separate wording for no attachments (wording to be improved later).
- Added beta tags to dropdown and modal.
- Moved modal to center of screen
- Resolved issue of dropdown disconnecting
- Added google analytics monitoring for failed downloads (Closes #762)

image

**Comment by liangyuanruo** [comment]
code looks good, @syan-syan if you could confirm that the wording is OK?

File: src/public/modules/forms/admin/views/download-all-attachments.client.modal.html [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        each response will be downloaded as a separate ZIP file. Start now, or
+        adjust your date range before proceeding.
+      </p>
+    </div>
+    <div ng-if="responsesCount === 0" class="field-description">
+      <p>
+        The date range you selected does not contain any responses. Please
+        select a date range containing responses and try again.
+      </p>
+    </div>
```
Comment (liangyuanruo) [inline src/public/modules/forms/admin/views/download-all-attachments.client.modal.html:19]: @syan-syan the wording is here

**Comment by liangyuanruo** [comment]
@frankchn I just had a quick discussion with @syan-syan, and we'd like to clear one last hurdle before releasing next week which is that the submission ID and timestamp aren't being included in the downloaded CSV if any part of the web worker decryption fails. It's important because, without it, the user can't provide us any info or contact us for help. 

could you help us to add that in? the other form fields can simply be left blank.

**Comment by frankchn** [comment]
Here are the current designs. They should be pretty close to the Zeplin references.

image

image

image

image

Also, added the submission ID and worker timestamp into the CSV as an additional column regardless of whether the submission succeeds or fails.

```
Expected total responses,2
Total decrypted responses,2
Error count,0
Unverified response count,0
Reference number,Timestamp,test attachment 1,test text 123,Download Status
5fd6a89fc9b363002aca6c6d,14 Dec 2020 07:49:51 AM,abcdef.pdf,STATUS123456,Success
5fd6a75fc9b363002aca6c69,14 Dec 2020 07:44:31 AM,ghijkl.jpg,,Success
```

**Comment by liangyuanruo** [comment]
thanks @frankchn ... i believe @syan-syan is envisioning a slightly different format for the CSV. let's firm it up at sync-up tomorrow.

**Comment by frankchn** [comment]
Changed CSV downloads to be the following format:

```
Expected total responses,2
Success count,2
Error count,0
Unverified response count,0
See download status column for download errors
Reference number,Timestamp,Attachment Download Status,test attachment 1,test text 123
5fd6a89fc9b363002aca6c6d,14 Dec 2020 07:49:51 AM,Success,abcdef.pdf,STATUS123456
5fd6a75fc9b363002aca6c69,14 Dec 2020 07:44:31 AM,Success,def.jpg,123456
```

**Comment by liangyuanruo** [comment]
> ```
> Expected total responses,2
> Success count,2
> Error count,0
> Unverified response count,0
> See download status column for download errors
> Reference number,Timestamp,Attachment Download Status,test attachment 1,test text 123
> 5fd6a89fc9b363002aca6c6d,14 Dec 2020 07:49:51 AM,Success,abcdef.pdf,STATUS123456
> 5fd6a75fc9b363002aca6c69,14 Dec 2020 07:44:31 AM,Success,def.jpg,123456
> ```

@frankchn not sure if @syan-syan told you, but could we change `Attachment Download Status` column to the more generic `Download Status` that would also show the status for decryption failure etc.?

File: src/public/modules/forms/services/submissions.client.factory.js [hunk 5]
```diff
@@ -297,7 +302,12 @@ function SubmissionsFactory(
               const { data } = event
               const { csvRecord, downloadBlob } = data
 
-              if (csvRecord.status === 'ERROR') {
+              if (csvRecord.status === 'ATTACHMENT_ERROR') {
+                attachmentErrorCount++
+                errorCount++
+                // accumulate dataset anyway, since the download is otherwise OK
```
Comment (liangyuanruo) [inline src/public/modules/forms/services/submissions.client.factory.js:308]: we also want to accumulate the submission ID and timestamp, if the decryption fails

File: src/public/modules/forms/helpers/decryption.worker.js [hunk 6]
```diff
@@ -150,10 +152,27 @@ async function decryptIntoCsv(data) {
...
+              attachmentDownloadUrls,
+              secretKey,
+            ),
+          )
+          downloadStatus.answer = 'Success'
+        } catch (error) {
+          csvRecord.status = 'ATTACHMENT_ERROR'
+          downloadStatus.answer = 'Failed'
+        }
+        csvRecord.submissionData.record.unshift(downloadStatus)
```
Comment (liangyuanruo) [inline src/public/modules/forms/helpers/decryption.worker.js:175]: is it possible that `record` is uninitialised at this point, causing the worker to crash?

File: src/public/modules/forms/services/submissions.client.factory.js (hunk 5 above)
Comment (frankchn) [reply src/public/modules/forms/services/submissions.client.factory.js]: Changed

File: src/public/modules/forms/helpers/decryption.worker.js (hunk 6 above)
Comment (frankchn) [reply src/public/modules/forms/helpers/decryption.worker.js]: True, resolved with 93d52fe

**Comment by frankchn** [comment]
Design changes:

https://www.figma.com/file/ta5tjjBqphML8LPRkmxD53/DesignMaster_v2?node-id=928%3A45

**Review by liangyuanruo — CHANGES_REQUESTED** [review]
Code is good; merge pending copy changes.

**Comment by frankchn** [comment]
image
image
image

**Comment by syan-syan** [comment]
@liangyuanruo just made a tiny commit changing copy

File: src/public/modules/forms/admin/views/view-responses.client.view.html [hunk 7]
```diff
@@ -105,11 +105,44 @@
...
+              </li>
+              <li role="menuitem">
+                <a
+                  ng-click="vm.confirmSubmissionCountsBeforeDownload()"
+                  style="cursor: pointer"
+                  >CSV and attachments <span class="beta-icon">beta</span></a
+                >
+              </li>
+            </ul>
+          </div>
```
Comment (liangyuanruo) [inline src/public/modules/forms/admin/views/view-responses.client.view.html:145]: we should probably refactor this HTML and associated CSS into a new `<dropdown-menu-button>` component for reuse.

**Review by liangyuanruo — APPROVED** [review]
lgtm, let's test on UAT later

**Comment by syan-syan** [comment]
copy changes

- pre-download modal infobox: "Downloading many attachments can be an intensive operation" remove the full stop after this line
- pre-download modal infobox: "ensure network connectivity is stable"
- all modals: header should say "responses and attachments"

**Pull request merged**